drivers/pwm/pwm_nrfx.c

```diff
@@ -6,23 +6,28 @@
 #include <nrfx_pwm.h>
 #include <drivers/pwm.h>
 #include <pm/device.h>
+#include <drivers/pinctrl.h>
+#include <soc.h>
 #include <hal/nrf_gpio.h>
 #include <stdbool.h>
 
 #define LOG_LEVEL CONFIG_PWM_LOG_LEVEL
 #include <logging/log.h>
 LOG_MODULE_REGISTER(pwm_nrfx);
 
-#define PWM_NRFX_CH_POLARITY_MASK       BIT(15)
-#define PWM_NRFX_CH_PULSE_CYCLES_MASK   BIT_MASK(15)
-#define PWM_NRFX_CH_VALUE_NORMAL        PWM_NRFX_CH_POLARITY_MASK
-#define PWM_NRFX_CH_VALUE_INVERTED      (0)
-#define PWM_NRFX_CH_PIN_MASK            ~NRFX_PWM_PIN_INVERTED
+#define PWM_NRFX_CH_POLARITY_MASK          BIT(15)
+#define PWM_NRFX_CH_PULSE_CYCLES_MASK      BIT_MASK(15)
+#define PWM_NRFX_CH_VALUE(value, inverted) \
+	(value | (inverted ? 0 : PWM_NRFX_CH_POLARITY_MASK))
 
 struct pwm_nrfx_config {
 	nrfx_pwm_t pwm;
 	nrfx_pwm_config_t initial_config;
 	nrf_pwm_sequence_t seq;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pcfg;
+#endif
+	uint8_t inverted_channels;
```

Comment:
note: would it be possible to specify inversion as a custom DT property? See for example https://github.com/zephyrproject-rtos/zephyr/pull/43454/files#diff-5acb59be30b12a3d7d272f696839477c0fd56181c7e89be92fbea3e5ebd1699bR78-R112

Reply:
I updated the PR to use a custom property for channel inversion in the pinctrl case. For the non-pinctrl case, the "chX-inverted" properties need to be maintained. There is one question, though, that I'm not quite satisfied with. I need the information about inverted channels in the driver so that the playback data for channels can be set appropriately. I do it by examining the pin output state set by pinctrl at runtime. Is there an easy way I could do it at buildtime using the devicetree information provided for pinctrl?

Reply:
After looking at this once more, I think that maybe it is not that bad in the current form. Setting the initial state of PWM outputs in pinctrl is better, because this should be done before the pins are configured as outputs (otherwise glitches can appear on the lines), and the driver cannot do it, as it does not know which pins are to be used until pinctrl connects those. And initialization of `inverted_channels` at runtime has the advantage that it will work well also with dynamic pinctrl (theoretically, one could want to use an inverted channel in the default configuration and non-inverted in an alternative one).